File: lca.hs

-- Lowest Common Ancestor on binary tree.

import Text.Printf

data Tree a = Empty | Node a (Tree a) (Tree a) deriving Show

--Create tree
myTree :: Tree Int
myTree = Node 1 
	    (Node 2 
	        (Node 4 Empty Empty)
	        (Node 5 Empty Empty)
	    )
	    (Node 3 
	        (Node 6 
	            (Node 10 Empty Empty) 
	            (Node 8 Empty Empty)
	        )
	        (Node 7 
	            (Node 9 Empty Empty) 
	            Empty
	        )
	    )

--Solve for common ancestor
lca :: Eq a => Tree a -> a -> a -> Either Bool a
lca Empty _ _ = Left False
lca (Node v tl tr) n1 n2 = 
    let l = lca tl n1 n2
        r = lca tr n1 n2
        onroot = v == n1 || v == n2
    in case (l, r, onroot) of
        (Right a  , _         , _    ) -> Right a
        (_        , Right a   , _    ) -> Right a
        (Left True, Left True , _    ) -> Right v
        (Left True, _         , True ) -> Right v
        (_        , Left True , True ) -> Right v
        (Left True, _         , False) -> Left True
        (_         , Left True, False) -> Left True
        (_         , _        , True ) -> Left True
        _ -> Left False

lca_show :: (Eq a, Show a, PrintfArg a) => Tree a -> a -> a -> String
lca_show t n1 n2 = printf "LCA(%d,%d)=%s" n1 n2 result
    where result = case lca t n1 n2 of
                        Right a -> show a
                        _ -> "not found"

--Test for examples
main = mapM print [
            lca_show myTree 0 9, --Expected: "not found"
            lca_show myTree 0 5, --Expected: "not found"
            lca_show myTree 4 5, --Expected: "2"
            lca_show myTree 4 6, --Expected: "1"
            lca_show myTree 3 4, --Expected: "1"
            lca_show myTree 2 4, --Expected: "2"
            lca_show myTree 9 10]--Expected: "3"

File: LCA.py

#Lowest common ancestor solution for a binary tree in python.

class Node:

	#Construct new node with a key and child nodes
	def __init__(self, key):
	    self.key = key
	    self.left = None
	    self.right = None


# Gets path from tree root to given node
# Stores path in path[]
# Returns false if no path found
def getPath(root, path, k):
	if root is None:
		return False

	path.append(root.key)

	if root.key == k:
		return True

	if((root.left != None and getPath(root.left, path, k)) or 
		(root.right != None and getPath(root.right, path, k))):
		return True
	
	path.pop()
	return False

# Returns lowest common ancestor of nodeA and nodeB
# Returns -1 if node isnt present in tree
def getLCA(root, nodeA, nodeB):
	path1 = []
	path2 = []

	if(not getPath(root, path1, nodeA) or not getPath(root, path2, nodeB)):
		return -1

	i=0
	while (i<len(path1) and i < len(path2)):
		if path1[i] != path2[i]:
			break
		i+=1

	return path1[i-1]


root = Node(1)
root.left = Node(2)
root.right = Node(4)
root.left.left = Node(3)
root.left.right = Node(11)
root.right.right = Node(6)
root.right.left = Node(7)
		